crates/apollo-compiler/src/name.rs

```diff
@@ -0,0 +1,511 @@
+use crate::diagnostic::CliReport;
+use crate::diagnostic::ToCliReport;
+use crate::execution::GraphQLLocation;
+use crate::node::TaggedFileId;
+use crate::schema::ComponentName;
+use crate::schema::ComponentOrigin;
+use crate::FileId;
+use crate::NodeLocation;
+use crate::SourceMap;
+use rowan::TextRange;
+use std::fmt;
+use std::marker::PhantomData;
+use std::mem::size_of;
+use std::mem::ManuallyDrop;
+use std::ptr::NonNull;
+use std::sync::Arc;
+
+/// Create a [`Name`] from a string literal or identifier, checked for validity at compile time.
+///
+/// A `Name` created this way does not own allocated heap memory or a reference counter,
+/// so cloning it is extremely cheap.
+///
+/// # Examples
+///
+/// ```
+/// use apollo_compiler::name;
+///
+/// assert_eq!(name!("Query").as_str(), "Query");
+/// assert_eq!(name!(Query).as_str(), "Query");
+/// ```
+///
+/// ```compile_fail
+/// # use apollo_compiler::name;
+/// // error[E0080]: evaluation of constant value failed
+/// // assertion failed: ::apollo_compiler::ast::Name::valid_syntax(\"è_é\")
+/// let invalid = name!("è_é");
+/// ```
+#[macro_export]
+macro_rules! name {
+    ($value: ident) => {
+        $crate::name!(stringify!($value))
+    };
+    ($value: expr) => {{
+        const _: () = { assert!($crate::Name::valid_syntax($value)) };
+        $crate::Name::new_static_unchecked(&$value)
+    }};
+}
+
+/// A GraphQL identifier
+///
+/// Like [`Node`][crate::Node], this string type has cheap `Clone`
+/// and carries an optional source location.
+///
+/// Internally, the string value is either an atomically-reference counter `Arc<str>`
+/// or a `&'static str` borrow that lives until the end of the program.
+//
+// Fields: equivalent to `(UnpackedRepr, Option<NodeLocation>)` but more compact
+pub struct Name {
+    /// Data pointer of either `Arc<str>` (if `tagged_file_id.tag() == TAG_ARC`)
+    /// or `&'static str` (if `TAG_STATIC`)
+    ptr: NonNull<u8>,
+    len: u32,
+    start_offset: u32,            // zero if we don’t have a location
+    tagged_file_id: TaggedFileId, // `.file_id() == FileId::NONE` means we don’t have a location
+    phantom: PhantomData<UnpackedRepr>,
+}
+
+#[allow(dead_code)] // only used in PhantomData and static asserts
+enum UnpackedRepr {
+    Heap(Arc<str>),
+    Static(&'static str),
+}
+
+#[derive(Clone, Eq, PartialEq, thiserror::Error)]
+#[error("`{name}` is not a valid GraphQL name")]
+pub struct InvalidNameError {
+    name: String,
+    location: Option<NodeLocation>,
+}
+
```

Comment:
Despite having "invalid" in its name this is not an error returned by `Schema::validate` or `ExecutableDocument::validate`, only by `Name::new` or other constructors.

We made `Name` a distinct Rust type so that `validate()` does not need to check name validity at all: https://github.com/apollographql/apollo-rs/issues/710